### PR TITLE
Re-enable custom op support

### PR DESCRIPTION
**Pull request by nithinsubbiah**
Re-enable custom op support (https://github.com/llvm/torch-mlir/pull/895) that was reverted due to MacOS builder issue.

### CODE REVIEW

**Comment by nithinsubbiah** [comment]
@rdadolf Python build fails because I haven't included the `__init__.py` that loads the custom_op_example library file (.so) - https://github.com/llvm/torch-mlir/pull/1139/files#diff-f7cdb8efffece9ed974b90846e54110d5d9e1b808a46225538e54e09772877b2L12. When I include it I get the following error: `torch-mlir/python/torch_mlir/_torch_mlir_custom_op_example/__init__.py): Too many levels of symbolic links`. My understanding is that a symbolic link to itself is created somewhere but I can't figure out where or how. If you do see something fishy please let me know, thanks!

File: python/torch_mlir/_torch_mlir_custom_op_example/CMakeLists.txt [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # Setup PyTorch
-list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/../dialects/torch/importer/jit_ir/cmake/modules")
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/../cmake/modules")
 include(TorchMLIRPyTorch)
 TorchMLIRProbeForPyTorchInstall()
```
Comment (powderluv) [inline python/torch_mlir/_torch_mlir_custom_op_example/CMakeLists.txt:4]: you want to check if it is a LOCAL Pytorch install

Comment (nithinsubbiah) [reply python/torch_mlir/_torch_mlir_custom_op_example/CMakeLists.txt:4]: I changed this PR to just enable custom op support, will modify this in the following PR which will have this example working

**Comment by nithinsubbiah** [comment]
I've modified this PR to just enable custom op support - ability to register custom ops as torch ops in Torch MLIR. I will enable the example in a follow-up PR (to better resolve the issue)

File: build_tools/update_shape_lib.sh [hunk 2]
```diff
@@ -9,26 +9,27 @@
 # For more information on supporting custom operators, see:
 #   ${TORCH_MLIR}/python/torch_mlir/_torch_mlir_custom_op_example/README.md
 
-set -euo pipefail
+set -eo pipefail
```
Comment (powderluv) [inline build_tools/update_shape_lib.sh:12]: why this change ?

File: build_tools/update_shape_lib.sh [hunk 3]
```diff
@@ -9,26 +9,27 @@
...
 
 pypath="${python_packages_dir}/torch_mlir"
-# TODO: Re-enable once custom op support is back.
-#if [ ! -z ${TORCH_MLIR_EXT_PYTHONPATH} ]; then
-#  pypath="${pypath}:${TORCH_MLIR_EXT_PYTHONPATH}"
-#fi
-#ext_module="torch_mlir._torch_mlir_custom_op_example"
-#if [ ! -z ${TORCH_MLIR_EXT_MODULES} ]; then
+if [ ! -z ${TORCH_MLIR_EXT_PYTHONPATH} ]; then
+ pypath="${pypath}:${TORCH_MLIR_EXT_PYTHONPATH}"
```
Comment (powderluv) [inline build_tools/update_shape_lib.sh:21]: two spaces

File: build_tools/update_torch_ods.sh [hunk 4]
```diff
@@ -9,27 +9,28 @@
 # For more information on supporting custom operators, see:
 #   ${TORCH_MLIR}/python/torch_mlir/_torch_mlir_custom_op_example/README.md
 
-set -euo pipefail
+set -eo pipefail
```
Comment (powderluv) [inline build_tools/update_torch_ods.sh:12]: why ? can we leave as is ?

File: build_tools/update_torch_ods.sh [hunk 5]
```diff
@@ -9,27 +9,28 @@
...
 
 pypath="${python_packages_dir}/torch_mlir"
-# TODO: Re-enable once custom op support is back.
-#if [ ! -z ${TORCH_MLIR_EXT_PYTHONPATH} ]; then
-#  pypath="${pypath}:${TORCH_MLIR_EXT_PYTHONPATH}"
-#fi
-#ext_module="torch_mlir._torch_mlir_custom_op_example"
-#if [ ! -z ${TORCH_MLIR_EXT_MODULES} ]; then
+if [ ! -z ${TORCH_MLIR_EXT_PYTHONPATH} ]; then
+ pypath="${pypath}:${TORCH_MLIR_EXT_PYTHONPATH}"
```
Comment (powderluv) [inline build_tools/update_torch_ods.sh:21]: two spaces

File: python/torch_mlir/_torch_mlir_custom_op_example/CMakeLists.txt [hunk 6]
```diff
@@ -29,4 +29,3 @@ set_target_properties(torch_mlir_custom_op_example PROPERTIES
 )
 torch_mlir_python_target_compile_options(torch_mlir_custom_op_example)
 mlir_check_all_link_libraries(torch_mlir_custom_op_example)
-
```
Comment (powderluv) [inline python/torch_mlir/_torch_mlir_custom_op_example/CMakeLists.txt:32]: white space

File: build_tools/update_shape_lib.sh (hunk 2 above)
Comment (nithinsubbiah) [reply build_tools/update_shape_lib.sh:12]: `set -u` throws an unbound variable error if a referenced variable isn't set. When a user doesn't want to add a custom operator `TORCH_MLIR_EXT_PYTHONPATH` and `TORCH_MLIR_EXT_MODULES` variables aren't initialized thus relaxing the need for it to be set.

File: build_tools/update_torch_ods.sh (hunk 4 above)
Comment (nithinsubbiah) [reply build_tools/update_torch_ods.sh:12]: same reasoning as above

File: build_tools/update_shape_lib.sh (hunk 2 above)
Comment (powderluv) [reply build_tools/update_shape_lib.sh:12]: I think we should set it to default if not set.

File: python/CMakeLists.txt [hunk 7]
```diff
@@ -111,8 +111,8 @@ add_subdirectory(torch_mlir/eager_mode)
 # Required for running the update_torch_ods.sh and update_shape_lib.sh scripts.
 ################################################################################
 
-# TODO: renable once it build on macOS Intel / M1
-#add_subdirectory(torch_mlir/_torch_mlir_custom_op_example)
+#TODO: reenable custom op example
+# add_subdirectory(torch_mlir/_torch_mlir_custom_op_example)
```
Comment (powderluv) [inline python/CMakeLists.txt:115]: i would also remove all these TODOs. I think @silvasean is going to rework custom ops in the future and these may not make sense.

File: build_tools/update_shape_lib.sh (hunk 2 above)
Comment (powderluv) [reply build_tools/update_shape_lib.sh:12]: like https://github.com/llvm/torch-mlir/blob/9cf0b6e8ff4ed2400175475c7ca7694688cf17f1/build_tools/python_deploy/build_linux_packages.sh#L43

File: lib/Conversion/TorchToLinalg/CMakeLists.txt [hunk 8]
```diff
@@ -1,6 +1,5 @@
 add_mlir_conversion_library(TorchMLIRTorchToLinalg
-# TODO: Re-enable after MacOS support is fixed for the custom op extension.
-#  CustomOpExample.cpp
+  # CustomOpExample.cpp
```
Comment (powderluv) [inline lib/Conversion/TorchToLinalg/CMakeLists.txt:2]: delete ?

File: lib/Conversion/TorchToLinalg/PopulatePatterns.h [hunk 9]
```diff
@@ -63,7 +63,7 @@ void populateIndirectDataMovementPatternsAndLegality(
 void populateTensorConstructorsPatternsAndLegality(TypeConverter &typeConverter,
                                                    RewritePatternSet &patterns,
                                                    ConversionTarget &target);
-//void populateCustomOpExamplePatternsAndLegality(TypeConverter &typeConverter,
+// void populateCustomOpExamplePatternsAndLegality(TypeConverter &typeConverter,
```
Comment (powderluv) [inline lib/Conversion/TorchToLinalg/PopulatePatterns.h:66]: whitespace

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp [hunk 10]
```diff
@@ -62,7 +62,7 @@ class ConvertTorchToLinalg
 
     RewritePatternSet patterns(context);
 
-    //torch_to_linalg::populateCustomOpExamplePatternsAndLegality(
+    // torch_to_linalg::populateCustomOpExamplePatternsAndLegality(
```
Comment (powderluv) [inline lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:65]: whitespace

**Review by powderluv — COMMENTED** [review]
some more whitespace changes

**Review by powderluv — APPROVED** [review]
